v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -82,37 +85,96 @@ func (c *armClientCache) lookup(key string) (*armClient, bool) {
 
 // GetClient finds and returns a client and credential to be used for a given resource
 func (c *armClientCache) GetClient(ctx context.Context, obj genruntime.ARMMetaObject) (*genericarmclient.GenericClient, string, error) {
+
+	client, err := c.checkPerResourceSecret(ctx, obj)
+	if err != nil {
+		return nil, "", err
+	} else if client != nil {
+		return client.GenericClient(), client.CredentialFrom(), nil
+	}
 	// Namespaced secret
-	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
+	client, err = c.checkNamespacedSecret(ctx, obj)
 	if err != nil {
 		return nil, "", err
+	} else if client != nil {
+		return client.GenericClient(), client.CredentialFrom(), nil
+	}
+
+	// If not found, default would be global
+	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+}
+
+func (c *armClientCache) checkNamespacedSecret(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
+	if err != nil {
+		return nil, err
 	}
 
 	if secret != nil {
-		nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
-		client, ok := c.lookup(nsName.String())
-		// if client does not already exist, or secret data is changed, we need to get new token.
-		if !ok || !matchSecretData(client.secretData, secret.Data) {
-			credential, subscriptionID, err := c.newCredentialFromSecret(secret, nsName)
-			if err != nil {
-				return nil, "", err
-			}
-
-			newClient, err := genericarmclient.NewGenericClientFromHTTPClient(c.cloudConfig, credential, c.httpClient, subscriptionID, metrics.NewARMClientMetrics())
-			if err != nil {
-				return nil, "", err
-			}
-
-			armClient := newARMClient(newClient, secret.Data, nsName)
-			c.register(armClient)
-			return armClient.GenericClient(), armClient.CredentialFrom(), nil
-		} else {
-			return client.GenericClient(), client.CredentialFrom(), nil
+		armClient, err := c.getARMClientFromSecret(secret)
+		if err != nil {
+			return nil, err
 		}
+		return armClient, nil
 	}
+	return nil, nil
+}
 
-	// If not found, default would be global
-	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+func (c *armClientCache) checkPerResourceSecret(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	credentialFrom, ok := obj.GetAnnotations()[perResourceSecretAnnotation]
+	// annotation exists, use per Resource Group secret credential
```

Comment:
minor:
```suggestion
	// annotation exists, use specificed secret
```

As this doesn't have to be a resourceGroup?